logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrator.kt

```diff
@@ -0,0 +1,98 @@
+import com.wire.kalium.logic.functional.getOrNull
+import com.wire.kalium.logic.kaliumLogger
+import com.wire.kalium.logic.wrapApiRequest
+import com.wire.kalium.network.api.base.authenticated.conversation.ConvProtocol
+import com.wire.kalium.network.api.base.authenticated.conversation.ConversationApi
+import com.wire.kalium.network.api.base.authenticated.conversation.UpdateConversationProtocolResponse
+import kotlinx.coroutines.flow.first
+
+interface MLSMigrator {
+    suspend fun migrateProteusConversations(): Either<CoreFailure, Unit>
+}
+
+class MLSMigratorImpl(
+    private val selfTeamIdProvider: SelfTeamIdProvider,
+    private val conversationRepository: ConversationRepository,
+    private val mlsConversationRepository: MLSConversationRepository,
+    private val conversationApi: ConversationApi
+) : MLSMigrator {
+
+    override suspend fun migrateProteusConversations(): Either<CoreFailure, Unit> =
+        conversationRepository.getConversationList().flatMap {
+            it.first().filter { conversation ->
+                conversation.protocol is Conversation.ProtocolInfo.Proteus &&
+                    conversation.type == Conversation.Type.GROUP &&
+                    conversation.isTeamGroup() &&
+                    conversation.teamId == selfTeamIdProvider().getOrNull()
+            }.foldToEitherWhileRight(Unit) { conversation, _ ->
+                migrate(conversation.id)
+            }
+        }
+
+    private suspend fun migrate(conversationId: ConversationId): Either<CoreFailure, Unit> {
+        kaliumLogger.i("migrating ${conversationId.toLogString()} to mixed")
+        return updateProtocolToMixed(conversationId)
+            .flatMap {
+                establishConversation(conversationId)
+            }.flatMapLeft {
+                kaliumLogger.w("failed to migrate ${conversationId.toLogString()} to mixed: $it")
+                Either.Right(Unit)
+            }
+    }
+
+    private suspend fun updateProtocolToMixed(conversationId: ConversationId) =
+        wrapApiRequest { conversationApi.updateProtocol(conversationId.toApi(), ConvProtocol.MIXED) }
+            .flatMap {
+                if (it is UpdateConversationProtocolResponse.ProtocolUpdated) {
+                    conversationRepository.fetchConversation(conversationId)
+                } else {
+                    Either.Right(Unit)
+                }
+            }
+
+    private suspend fun establishConversation(conversationId: ConversationId) =
+        conversationRepository.getConversationProtocolInfo(conversationId)
+            .flatMap { protocolInfo ->
+                when (protocolInfo) {
+                    is Conversation.ProtocolInfo.MLS -> { // TODO jacob should be MIXED
+                        mlsConversationRepository.establishMLSGroup(protocolInfo.groupId, emptyList())
+                            .flatMap {
+                                conversationRepository.getConversationMembers(conversationId).flatMap { members ->
```

Comment:
what is our recovery strategy if one of these steps fail?

Reply:
Try again in 24 hours